flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesRunningJobsRegistry.java

```diff
@@ -0,0 +1,136 @@
+ */
+
+package org.apache.flink.kubernetes.highavailability;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.kubernetes.kubeclient.FlinkKubeClient;
+import org.apache.flink.kubernetes.kubeclient.resources.KubernetesException;
+import org.apache.flink.kubernetes.kubeclient.resources.KubernetesLeaderElector;
+import org.apache.flink.runtime.highavailability.RunningJobsRegistry;
+import org.apache.flink.util.StringUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static org.apache.flink.kubernetes.utils.Constants.RUNNING_JOBS_REGISTRY_KEY_PREFIX;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link RunningJobsRegistry} implementation for Kubernetes. All the running jobs will be stored in
+ * Dispatcher-leader ConfigMap. The key is the job id with prefix
+ * {@link org.apache.flink.kubernetes.utils.Constants#RUNNING_JOBS_REGISTRY_KEY_PREFIX},
+ * and value is job status.
+ */
+public class KubernetesRunningJobsRegistry implements RunningJobsRegistry {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KubernetesRunningJobsRegistry.class);
+
+	private final FlinkKubeClient kubeClient;
+
+	private final String configMapName;
+
+	private final String lockIdentity;
+
+	public KubernetesRunningJobsRegistry(FlinkKubeClient kubeClient, String configMapName, String lockIdentity) {
+		this.kubeClient = checkNotNull(kubeClient);
+		this.configMapName = checkNotNull(configMapName);
+		this.lockIdentity = checkNotNull(lockIdentity);
+	}
+
+	@Override
+	public void setJobRunning(JobID jobID) throws IOException {
+		checkNotNull(jobID);
+
+		writeJobStatusToConfigMap(jobID, JobSchedulingStatus.RUNNING);
+	}
+
+	@Override
+	public void setJobFinished(JobID jobID) throws IOException {
+		checkNotNull(jobID);
+
+		writeJobStatusToConfigMap(jobID, JobSchedulingStatus.DONE);
+	}
+
+	@Override
+	public JobSchedulingStatus getJobSchedulingStatus(JobID jobID) throws IOException {
+		checkNotNull(jobID);
+
+		return kubeClient.getConfigMap(configMapName)
+			.map(configMap -> {
+				final String key = getKeyForJobId(jobID);
+				final String status = configMap.getData().get(key);
+				if (!StringUtils.isNullOrWhitespaceOnly(status)) {
+					return JobSchedulingStatus.valueOf(status);
+				} else {
+					LOG.warn("ConfigMap {} contained corrupted data. Ignoring the key {}.", configMapName, key);
+				}
+				return JobSchedulingStatus.PENDING;
+			})
+			.orElseThrow(() -> new IOException(
+				new KubernetesException("ConfigMap " + configMapName + " does not exist.")));
+	}
+
+	@Override
+	public void clearJob(JobID jobID) throws IOException {
+		checkNotNull(jobID);
+
+		try {
+			kubeClient.checkAndUpdateConfigMap(
+				configMapName,
+				configMap -> {
+					if (KubernetesLeaderElector.hasLeadership(configMap, lockIdentity)) {
+						configMap.getData().remove(getKeyForJobId(jobID));
+						return Optional.of(configMap);
```

Comment:
Might be better to check the original value and try to avoid the update if the value is not changed.